# Funcion Examples

### Example 1. Temperature convertion

Celsius to Fahrenheit:
F = (9/5) * C + 32

Fahrenheit to Celsius:
C = (5/9) * (F - 32)


In [ ]:
## To convert the temperature from Celsius to Fahrenheit, we need 2 variable.
## 1st variable = Unit of temperature (Celsius or Fahrenheit)
## 2nd variable = Value of temperature (numerical value)

In [24]:
def convert_temperature (unit,value):
    if unit == "c":
        return value * (9/5) + 32   # formula to convert Celsius to Fahrenheit

    elif unit == "f":
        return value * (5/9) - 32   # formula to convert Fahrenheit to Celsius

    else:
        return "Invalid unit of temperature. Please use 'c' for Celsius or 'f' for Fahrenheit."


In [25]:
## calling the function 
print(convert_temperature("f",32))
print(convert_temperature("c",100))


-14.222222222222221
212.0


### Exaple 2. Password Strength Checker

#### Strong Password Default Parameters

- Minimum length: 8–12 characters (recommended 12+)
- At least 1 uppercase letter (A–Z)
- At least 1 lowercase letter (a–z)
- At least 1 number (0–9)
- At least 1 special character (!@#$%^&*)
- Should not contain common words or personal information

#### Example Strong Password

```
SwaP@2026#Data
```


In [33]:
def is_strong_password(password):
    """ This function checks if the password is strong or not"""
    if len(password)<8 :
        return False
    if not any(char.isdigit() for char in password):
        return False
    if not any(char.isupper() for char in password):
        return False
    if not any(char.islower() for char in password):
        return False
    if not any(char in '!@#$%^&*' for char in password):
        return False
    else:
        return True

        

In [34]:
print(is_strong_password("Abc123!@#"))  # True
print(is_strong_password("abc123"))      # False (no uppercase and special character)

True
False


In [30]:
password = input("Enter your password: ")

# has_upper = False
# has_lower = False
# has_digit = False
# has_special = False

# special_chars = "!@#$%^&*"

# Loop through each character
for char in password:
    if char.isupper():
        has_upper = True
    elif char.islower():
        has_lower = True
    elif char.isdigit():
        has_digit = True
    elif char in special_chars:
        has_special = True

# Checking conditions using if-else
if len(password) < 8:
    print("Password must be at least 8 characters long.")

elif not has_upper:
    print("Password must contain at least one uppercase letter.")

elif not has_lower:
    print("Password must contain at least one lowercase letter.")

elif not has_digit:
    print("Password must contain at least one number.")

elif not has_special:
    print("Password must contain at least one special character (!@#$%^&*).")

else:
    print("Strong Password ✅")


Password must be at least 8 characters long.


### Example 3. Calculate the Total Cost of item in shopping cart

In [41]:
def calculate_total_cost(cart):
    total_cost=0
    for item in cart:
        total_cost+=item["Price"]* item["quantity"]
        return total_cost

## Example cart data

cart=[
    {"name":"apple",'Price':0.5,"quantity":4},
    {"name":"Banana",'Price':0.3,"quantity":6},
    {"name":"Orange",'Price':0.7,"quantity":3}
    ]
## calling the funciton 
total_cost=calculate_total_cost(cart)
print(f"Total cost of the cart is: ${total_cost}")

Total cost of the cart is: $2.0


### Example 4: check IF a String is palindrome

##### Palindrome

A **palindrome** is a word, number, or sentence that reads the same forward and backward.

#### Examples

#### Words
- madam
- level
- radar
- racecar

#### Numbers
- 121
- 1331
- 12321

#### Sentence
- Madam
- Racecar

If the original value is equal to its reversed value, then it is called a **palindrome**.


In [46]:
def is_palindrome(s):
    s=s.lower().replace(" ","")
    return s==s[::-1]

print(is_palindrome("madam"))
print(is_palindrome("Hello"))

True
False


### Example 5. Calculate the factorial of a number using recursion

Recursion

Recursion is a programming technique where a function calls itself to solve a smaller part of the same problem.

In simple words:

A function that calls itself is called recursion.

In [61]:
def factorial(n):
    if n==0:
        return 1    ## Base case: factorial of 0 is 1
    else:
        return n * factorial(n-1)     ## Recursive case: n! = n * (n-1)!

print(factorial(5))

120


# 📘 Recursion Complete Notes (Factorial Example)

---

## 🔁 What is Recursion?

Recursion is a programming technique where a function **calls itself** to solve a smaller part of the same problem.

Every recursive function must have:

1. ✅ Base Case (Stopping condition)
2. 🔄 Recursive Case (Function calls itself)

---

## 🧮 Example: Factorial

Mathematical Definition:

n! = n × (n-1)!

Example:
5! = 5 × 4 × 3 × 2 × 1 = 120

---

## 🧑‍💻 Python Code

```python
def factorial(n):
    if n == 0:          # Base Case
        return 1
    else:               # Recursive Case
        return n * factorial(n - 1)

print(factorial(5))
```

---

# 📥 Phase 1: Stack Building (Going Down)

When we call:

```
factorial(5)
```

Each function call waits for the next one.

| Step | Function Call | Waiting For        | Stack Status |
|------|--------------|-------------------|-------------|
| 1    | factorial(5) | factorial(4)      | Waiting     |
| 2    | factorial(4) | factorial(3)      | Waiting     |
| 3    | factorial(3) | factorial(2)      | Waiting     |
| 4    | factorial(2) | factorial(1)      | Waiting     |
| 5    | factorial(1) | factorial(0)      | Waiting     |
| 6    | factorial(0) | —                 | Returns 1   |

At Step 6, the **Base Case is reached**.

---

# 📤 Phase 2: Stack Resolving (Coming Back)

Now values return upward step-by-step.

| Step | Returning To   | Calculation     | Result |
|------|---------------|----------------|--------|
| 1    | factorial(1)  | 1 × 1          | 1      |
| 2    | factorial(2)  | 2 × 1          | 2      |
| 3    | factorial(3)  | 3 × 2          | 6      |
| 4    | factorial(4)  | 4 × 6          | 24     |
| 5    | factorial(5)  | 5 × 24         | 120    |

---

# 🎯 Final Output

```
120
```

---

# 🧠 Important Concept

Recursion works in two phases:

1. 📥 Going Down (Stack Building)
2. 📤 Coming Back (Stack Resolving)

Each function call waits for the next one to finish.

---

# ⚠ Why Base Case is Important

If we remove the base case:

```python
def factorial(n):
    return n * factorial(n - 1)
```

The function will call itself forever.

This causes:

```
RecursionError: maximum recursion depth exceeded
```

So, a base case is necessary to stop recursion.

---

# 🪜 Simple Mental Model

Think of recursion like climbing stairs:

Going Down:
5 → 4 → 3 → 2 → 1 → 0  

Coming Back:
1 → 2 → 6 → 24 → 120  

---

# ✅ Key Takeaway

Recursion = Function calling itself  
Base Case = Stops recursion  
Recursive Case = Breaks problem into smaller part  
Stack = Stores function calls in memory


# Recursion Stack Table (Factorial Example)

## Function

```python
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n - 1)
```

---

## 📥 Phase 1: Stack Building (Going Down)

| Step | Function Call   | Status                     |
|------|-----------------|----------------------------|
| 1    | factorial(5)    | Waiting for factorial(4)   |
| 2    | factorial(4)    | Waiting for factorial(3)   |
| 3    | factorial(3)    | Waiting for factorial(2)   |
| 4    | factorial(2)    | Waiting for factorial(1)   |
| 5    | factorial(1)    | Waiting for factorial(0)   |
| 6    | factorial(0)    | Base Case → Returns 1      |

---

## 📤 Phase 2: Stack Resolving (Coming Back)

| Step | Returning To     | Calculation         | Result |
|------|------------------|--------------------|--------|
| 1    | factorial(1)     | 1 × 1              | 1      |
| 2    | factorial(2)     | 2 × 1              | 2      |
| 3    | factorial(3)     | 3 × 2              | 6      |
| 4    | factorial(4)     | 4 × 6              | 24     |
| 5    | factorial(5)     | 5 × 24             | 120    |

---

## ✅ Final Output

120


### Example 6. A function to read a file and count the frequency of each word

In [75]:
def count_word_frequency(file_path):
    word_count = {}

    with open(file_path, 'r') as file:
        for line in file:
            words = line.split()

            for word in words:
                word = word.lower().strip('!@#$%^&*.,:;')
                word_count[word] = word_count.get(word, 0) + 1

    return word_count


file_path = 'sample.txt'
frequency = count_word_frequency(file_path)
print(frequency)


{'recursion': 3, 'is': 3, 'a': 9, 'programming': 1, 'technique': 1, 'in': 3, 'which': 1, 'function': 4, 'calls': 2, 'itself': 2, 'to': 3, 'solve': 1, 'smaller': 2, 'version': 1, 'of': 2, 'the': 6, 'same': 1, 'problem': 2, 'it': 2, 'works': 1, 'by': 2, 'breaking': 1, 'large': 1, 'into': 1, 'sub-problems': 1, 'until': 1, 'stopping': 1, 'condition': 1, 'called': 1, 'base': 3, 'case': 3, 'reached': 1, 'once': 1, 'returns': 1, 'value': 1, 'start': 1, 'returning': 1, 'one': 2, 'reverse': 1, 'order': 1, 'combining': 1, 'their': 1, 'results': 1, 'produce': 1, 'final': 1, 'answer': 1, 'uses': 1, 'stack': 1, 'memory': 1, 'keep': 1, 'track': 1, 'each': 1, 'call': 1, 'very': 1, 'important': 1, 'because': 1, 'without': 1, 'would': 1, 'continue': 1, 'calling': 1, 'forever': 1, 'and': 1, 'cause': 1, 'error': 1}


### Example 7: Validate email address

In [77]:
import re

# Function to validate email address
def is_valid_email(email):
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return re.match(pattern, email) is not None 

# calling the function  
print(is_valid_email("test@example.com"))  # Should return True
print(is_valid_email("invalid-email"))     # Should return False

True
False
